src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -343,26 +395,58 @@ private bool IsRecovering(ServerErrorCode code, string message)
             return false;
         }
 
+        private bool IsRecoveringErrorException(Exception exception)
+        {
+            return
+                exception is MongoCommandException commandException &&
+                IsRecovering((ServerErrorCode)commandException.Code, commandException.ErrorMessage);
+        }
+
+        private void SetDescription(ServerDescription newDescription)
+        {
+            var oldDescription = Interlocked.CompareExchange(ref _currentDescription, null, null);
+            SetDescription(oldDescription, newDescription);
+        }
+
+        private void SetDescription(ServerDescription oldDescription, ServerDescription newDescription)
+        {
+            Interlocked.Exchange(ref _currentDescription, newDescription);
+            OnDescriptionChanged(sender: this, new ServerDescriptionChangedEventArgs(oldDescription, newDescription));
+        }
+
         private bool ShouldClearConnectionPoolForChannelException(Exception ex, SemanticVersion serverVersion)
         {
-            if (ex is MongoNotPrimaryException mongoNotPrimaryException && mongoNotPrimaryException.Code == (int)ServerErrorCode.NotMaster)
+            if (ex is MongoConnectionException mongoCommandException &&
+                mongoCommandException.IsNetworkException &&
+                !mongoCommandException.ContainsSocketTimeoutException)
             {
-                return !Feature.KeepConnectionPoolWhenNotMasterConnectionException.IsSupported(serverVersion);
+                return true;
```

Comment:
This is the missing bit of logic from the CSHARP-2126 PR: https://github.com/vincentkam/mongo-csharp-driver/pull/114:
From the [spec commit associated with that 2126 combined with the pre-existing language](https://github.com/mongodb/specifications/commit/bfddbd98bf396104f78500f93f1f13e3fd2d8c3a#diff-62b4da2e38e7d2a9c1a2d92d315895bbR1260-R1264)  (emphasis mine)
>  If, however, there is some other network error on the connection after the handshake completes, the client MUST replace the server's description with a default ServerDescription of type Unknown, and fill the ServerDescription's error field with useful information, the same as if an error or timeout occurred before the handshake completed.

When the client marks a server Unknown due to a network error or timeout, the Unknown ServerDescription MUST be sent through the same process for updating the TopologyDescription as if it had been a failed ismaster outcome from a server check: for example, if the TopologyType is ReplicaSetWithPrimary and a write to the RSPrimary server fails because of a network error (other than timeout), then a new ServerDescription is created for the primary, with type Unknown, and the client executes the proper subroutine for an Unknown server when the TopologyType is ReplicaSetWithPrimary: referring to the table above we see the subroutine is checkIfHasPrimary. The result is the TopologyType changes to ReplicaSetNoPrimary. See the test scenario called "Network error writing to primary".

> **The client MUST close all idle sockets in its connection pool for the server: if one socket is bad, it is likely that all are.**